{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Imports ###\n",
    "import os\n",
    "from typing import List, Tuple, Union, Dict\n",
    "from classes.client import Client\n",
    "from classes.coordinator_utils import select_common_features_variables, \\\n",
    "    compute_beta, create_beta_mask\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import time"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Helper Functions ###\n",
    "### Just run this, these functions are needed in various places ###\n",
    "\n",
    "### Define the client class ###\n",
    "class ClientWrapper:\n",
    "    \"\"\"\n",
    "    Holds all information necessary for the simulation run to work.\n",
    "    Defines the input data of the client, it's name and if it should be\n",
    "    considered as the coordinator\n",
    "    \"\"\"\n",
    "    def __init__(self, id: str, input_folder: str, coordinator: bool = False):\n",
    "        self.id = id\n",
    "        self.input_folder = input_folder\n",
    "        self.is_coordinator = coordinator\n",
    "        self.client_class: Client # initiated later\n",
    "        self.data_corrected: pd.DataFrame # initiated later\n",
    "\n",
    "def _check_consistency_clientwrappers(clientWrappers: List[ClientWrapper]) -> None:\n",
    "    \"\"\"\n",
    "    Checks for a list of clients if they were created correctly\n",
    "    Raises a ValueError in case of inconsistencies\n",
    "    Checks:\n",
    "        1. If exactly one coordinator exists\n",
    "    \"\"\"\n",
    "    coord = False\n",
    "    for clientWrapper in clientWrappers:\n",
    "        if coord and clientWrapper.is_coordinator:\n",
    "            raise ValueError(\"More than one coordinator was defined, please check \"+\\\n",
    "                            \"the code defining the clients\")\n",
    "        if clientWrapper.is_coordinator:\n",
    "            coord = True\n",
    "    if not coord:\n",
    "        raise ValueError(\"No client instance is a coordinator, please designate \"+\\\n",
    "                        \"any client as a coordinator\")\n",
    "\n",
    "def _compare_central_federated_dfs(name:str,\n",
    "                                   central_df: pd.DataFrame,\n",
    "                                   federated_df: pd.DataFrame,\n",
    "                                   intersection_features: List[str]) -> None:\n",
    "    \"\"\"\n",
    "    Compares two dataframes for equality. First checks that index and columns\n",
    "    are the same, then analyses the element wise differences.\n",
    "    See the analyse_diff_df function for more details on the difference analysis.\n",
    "    If both dataframes contain a NaN value at the same position, this is considered\n",
    "    as equal (0 as difference).\n",
    "    Args:\n",
    "        name: Name used for printing\n",
    "        central_df: The central dataframe\n",
    "        federated_df: The federated dataframe\n",
    "        intersection_features: The features that are common to both dataframes\n",
    "    \"\"\"\n",
    "    central_df = central_df.sort_index(axis=0).sort_index(axis=1)\n",
    "    federated_df = federated_df.sort_index(axis=0).sort_index(axis=1)\n",
    "    print(f\"_________________________Analysing: {name}_________________________\")\n",
    "    ### compare columns and index ###\n",
    "    failed = False\n",
    "    if not central_df.columns.equals(federated_df.columns):\n",
    "        print(f\"Columns do not match for central_df and federated_df\")\n",
    "        union_cols = central_df.columns.union(federated_df.columns)\n",
    "        intercept_cols = central_df.columns.intersection(federated_df.columns)\n",
    "        print(f\"Union-Intercept of columns: {union_cols.difference(intercept_cols)}\")\n",
    "        failed = True\n",
    "    if not central_df.index.equals(federated_df.index):\n",
    "        print(f\"Rows do not match for central_df and federated_df\")\n",
    "        union_rows = central_df.index.union(federated_df.index)\n",
    "        intercept_rows = central_df.index.intersection(federated_df.index)\n",
    "        print(f\"Union-Intercept of rows: {union_rows.difference(intercept_rows)}\")\n",
    "        failed = True\n",
    "    if failed:\n",
    "        print(f\"_________________________FAILED: {name}_________________________\")\n",
    "\n",
    "    df_diff = (central_df - federated_df).abs()\n",
    "    print(f\"Max difference: {df_diff.max().max()}\")\n",
    "    print(f\"Mean difference: {df_diff.mean().mean()}\")\n",
    "    print(f\"Max diff at position: {df_diff.idxmax().idxmax()}\")\n",
    "\n",
    "    df_diff_intersect = df_diff.loc[intersection_features]\n",
    "    print(f\"Max difference in intersect: {df_diff_intersect.max().max()}\")\n",
    "    print(f\"Mean difference in intersect: {df_diff_intersect.mean().mean()}\")\n",
    "    print(f\"Max diff at position in intersect: {df_diff_intersect.idxmax().idxmax()}\")\n",
    "\n",
    "def _concat_federated_results(clientWrappers: List[ClientWrapper],\n",
    "                              samples_in_columns=True) -> Tuple[pd.DataFrame, List[str]]:\n",
    "    \"\"\"\n",
    "    Concatenates the results of the federated clients into one dataframe\n",
    "    Also checks which features are common to all clients\n",
    "    and returns them\n",
    "    Args:\n",
    "        clientWrappers: List of ClientWrapper instances, containing the data\n",
    "            in the data_corrected attribute\n",
    "        samples_in_columns: If True, the samples are in the columns, if False\n",
    "            they are in the rows. For expression files this is true.\n",
    "            This decides how to aggregate the dataframes\n",
    "    Returns:\n",
    "        merged_df: The merged dataframe containing the data of all clients\n",
    "        intersection_features: The features that are common to all clients\n",
    "    \"\"\"\n",
    "    merged_df = None\n",
    "    intersection_features = set()\n",
    "    for clientWrapper in clientWrappers:\n",
    "        # get the data in the correct format\n",
    "        if not hasattr(clientWrapper, \"data_corrected\") or \\\n",
    "            clientWrapper.data_corrected is None:\n",
    "            raise ValueError(\"No data was found in the clientWrappers\")\n",
    "        corrected_data = clientWrapper.data_corrected\n",
    "        if not samples_in_columns:\n",
    "            corrected_data = corrected_data.T\n",
    "\n",
    "        cleaned_corrected_features = set(corrected_data.dropna().index)\n",
    "        # initialize the merged_df\n",
    "        if merged_df is None:\n",
    "            merged_df = corrected_data\n",
    "            intersection_features = cleaned_corrected_features\n",
    "            continue\n",
    "\n",
    "        # merge the data\n",
    "        merged_df = pd.concat([merged_df, corrected_data], axis=1)\n",
    "        intersection_features = intersection_features.intersection(cleaned_corrected_features)\n",
    "\n",
    "    # final check\n",
    "    if merged_df is None:\n",
    "        raise ValueError(\"No data was found in the clientWrappers\")\n",
    "    # reverse the Transpose if necessary\n",
    "    if not samples_in_columns:\n",
    "        merged_df = merged_df.T\n",
    "    return merged_df, list(intersection_features)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## DATA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### This part defines the data used. A ClientWrapper class is used to      ###\n",
    "### describe all cohorts.                                                  ###\n",
    "### Comment in the wanted data or add a new block for new data. No other   ###\n",
    "### changes are necessary.                                                 ###"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#################### Microarray Data ####################\n",
    "# First define the basefolder where all files are located\n",
    "base_dir = os.path.join(\"..\")\n",
    "# Go back to the git repos root dir\n",
    "datafolder = \"microarray\"\n",
    "base_dir = os.path.join(base_dir, \"evaluation_data\", \"microarray\", \"before\")\n",
    "\n",
    "# List of cohort names\n",
    "cohort_names = [\n",
    "    'GSE38666',  # Client 1 (Coordinator)\n",
    "    'GSE14407',  # Client 2\n",
    "    'GSE6008',   # Client 3\n",
    "    'GSE40595',  # Client 4\n",
    "    'GSE26712',  # Client 5\n",
    "    'GSE69428',  # Client 6\n",
    "]\n",
    "output_name = \"microarray\"\n",
    "central_filename = \"central_corrected_UNION.tsv\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# #################### Proteomics Data ####################\n",
    "\n",
    "# # First define the basefolder where all files are located\n",
    "# base_dir = os.path.join(\"..\")\n",
    "# # Go back to the git repos root dir\n",
    "# datafolder = \"proteomics\"\n",
    "# base_dir = os.path.join(base_dir, \"evaluation_data\", datafolder, \"before\")\n",
    "\n",
    "# # List of cohort names\n",
    "# cohort_names = [\n",
    "#     'lab_A',  # Client 1 (Coordinator)\n",
    "#     'lab_B',  # Client 2\n",
    "#     'lab_C',  # Client 3\n",
    "#     'lab_D',  # Client 4\n",
    "#     'lab_E',  # Client 5\n",
    "# ]\n",
    "# output_name = \"proteomics\"\n",
    "# central_filename = \"intensities_log_Rcorrected_UNION.tsv\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# #################### Proteomics Multibatch Data ####################\n",
    "\n",
    "# # First define the basefolder where all files are located\n",
    "# base_dir = os.path.join(\"..\")\n",
    "# # Go back to the git repos root dir\n",
    "# datafolder = \"proteomics_multibatch\"\n",
    "# base_dir = os.path.join(base_dir, \"evaluation_data\", datafolder, \"before\")\n",
    "\n",
    "# # List of cohort names\n",
    "# cohort_names = [\n",
    "#     'center1', # Client 1 (Coordinator)\n",
    "#     'center2', # Client 2\n",
    "#     'center3' # Client 3\n",
    "# ]\n",
    "# output_name = \"proteomics_multibatch\"\n",
    "# central_filename = \"intensities_log_Rcorrected_UNION.tsv\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# ################# Microbiome Data ####################\n",
    "# # First define the basefolder where all files are located\n",
    "# base_dir = os.path.join(\"..\")\n",
    "# # Go back to the git repos root dir\n",
    "# datafolder = \"microbiome\"\n",
    "# base_dir = os.path.join(base_dir, \"evaluation_data\", datafolder, \"before\")\n",
    "\n",
    "# # List of cohort names\n",
    "# cohort_names = [\n",
    "#     'PRJEB27928',  # Client 1 (Coordinator)\n",
    "#     'PRJEB6070',   # Client 2\n",
    "#     'PRJNA429097', # Client 3\n",
    "#     'PRJEB10878',  # Client 4\n",
    "#     'PRJNA731589', # Client 5\n",
    "# ]\n",
    "# output_name = \"microbiome\"\n",
    "# central_filename = \"normalized_logmin_counts_5centers_Rcorrected.tsv\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "# ################### Simulation Data ###################\n",
    "\n",
    "# # First define the basefolder where all files are located\n",
    "# base_dir = os.path.join(\"..\")\n",
    "# # Go back to the git repos root dir\n",
    "\n",
    "# # There are three different setups, activate the wanted one here accordingly, deactivate the others:\n",
    "# # BALANCED\n",
    "# # datafolder = os.path.join(\"simulated\", \"balanced\")\n",
    "# # base_dir = os.path.join(base_dir, \"evaluation_data\", \"simulated\", \"balanced\", \"before\")\n",
    "# # output_name = \"simulated_balanced\"\n",
    "# # # MILD IMBALANCED\n",
    "# # datafolder = os.path.join(\"simulated\", \"mild_imbalanced\")\n",
    "# # base_dir = os.path.join(base_dir, \"evaluation_data\", \"simulated\", \"mild_imbalanced\", \"before\")\n",
    "# # output_name = \"simulated_mild_imbalanced\"\n",
    "# # # STRONG IMBALANCED\n",
    "# datafolder = os.path.join(\"simulated\", \"strong_imbalanced\")\n",
    "# base_dir = os.path.join(base_dir, \"evaluation_data\", \"simulated\", \"strong_imbalanced\", \"before\")\n",
    "# output_name = \"simulated_strong_imbalanced\"\n",
    "\n",
    "# # List of cohort names\n",
    "# cohort_names = [\n",
    "#     'lab1',  # Client 1 (Coordinator)\n",
    "#     'lab2',  # Client 2\n",
    "#     'lab3',  # Client 3\n",
    "# ]\n",
    "# central_filename = \"intensities_R_corrected.tsv\" # is the same for all simulated setups"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# Initialize clientWrappers list\n",
    "clientWrappers: List[ClientWrapper] = []\n",
    "\n",
    "# Iterate over cohort names and create ClientWrapper instances\n",
    "for i, cohortname in enumerate(cohort_names):\n",
    "    clientWrappers.append(ClientWrapper(\n",
    "        id=cohortname,\n",
    "        input_folder=os.path.join(base_dir, cohortname),\n",
    "        coordinator=(i == 0)  # Set the first client as coordinator\n",
    "    ))\n",
    "\n",
    "# Double check that we only have one coordinator\n",
    "_check_consistency_clientwrappers(clientWrappers)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "# measure time for all clients\n",
    "time_tracker = {}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "###                                  INFO                                  ###\n",
    "### The following code blocks run the simulation. They are divided into    ###\n",
    "### multiple logical blocks to ease the use                                ###"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Got the following config:\n",
      "{'flimmaBatchCorrection': {'batch_col': 'batch', 'covariates': ['Pyr'], 'data_filename': 'intensities_log_UNION.tsv', 'design_filename': 'design.tsv', 'design_separator': '\\t', 'expression_file_flag': True, 'index_col': 'rowname', 'min_samples': 0, 'normalizationMethod': None, 'position': 0, 'separator': '\\t', 'smpc': False}}\n",
      "Opening dataset ../evaluation_data/proteomics_multibatch/before/center1/intensities_log_UNION.tsv\n",
      "Shape of rawdata(expr_file): (2910, 47)\n",
      "finished loading data, shape of data: (2910, 47), num_features: 2910, num_samples: 47\n",
      "Checking batch:  lab_A\n",
      "self.design has these batch values: ['lab_A' 'lab_B']\n",
      "Got these batch indices: ['Ref8537_QC1_20230414_1', 'Ref8537_QC2_20230414_1', 'Ref8537_QC3_20230414_1', 'Ref8537_QC4_20230414_1', 'Ref8537_S4_20230414', 'Ref8537_S8_20230414', 'Ref8537_S11_20230414', 'Ref8537_S18_20230414', 'Ref8537_S21_20230414', 'Ref8537_S26_20230414', 'Ref8537_S35_20230414', 'Ref8537_S37_20230414', 'Ref8537_S45_20230414', 'Ref8537_S49_20230414', 'Ref8537_S52_20230414', 'Ref8537_S60_20230414', 'Ref8537_S65_20230414', 'Ref8537_S69_20230414', 'Ref8537_S75_20230414', 'Ref8537_S79_20230414', 'Ref8537_S82_20230414', 'Ref8537_S88_20230414', 'Ref8537_S94_20230414', 'Ref8537_S100_20230414']\n",
      "Trying to access data ((2910, 47))\n",
      "Data has these cols: Index(['Ref8537_QC1_20230414_1', 'Ref8537_QC2_20230414_1',\n",
      "       'Ref8537_QC3_20230414_1', 'Ref8537_QC4_20230414_1',\n",
      "       'Ref8537_S4_20230414', 'Ref8537_S8_20230414', 'Ref8537_S11_20230414',\n",
      "       'Ref8537_S18_20230414', 'Ref8537_S21_20230414', 'Ref8537_S26_20230414',\n",
      "       'Ref8537_S35_20230414', 'Ref8537_S37_20230414', 'Ref8537_S45_20230414',\n",
      "       'Ref8537_S49_20230414', 'Ref8537_S52_20230414', 'Ref8537_S60_20230414',\n",
      "       'Ref8537_S65_20230414', 'Ref8537_S69_20230414', 'Ref8537_S75_20230414',\n",
      "       'Ref8537_S79_20230414', 'Ref8537_S82_20230414', 'Ref8537_S88_20230414',\n",
      "       'Ref8537_S94_20230414', 'Ref8537_S100_20230414',\n",
      "       'Clinspect_E_coli_A_QC1_Slot1-6_1_8640',\n",
      "       'Clinspect_E_coli_A_QC2_Slot1-15_1_8649',\n",
      "       'Clinspect_E_coli_A_S5_Slot1-7_1_8641',\n",
      "       'Clinspect_E_coli_A_S7_Slot1-24_1_8673',\n",
      "       'Clinspect_E_coli_A_S14_Slot1-11_1_8645',\n",
      "       'Clinspect_E_coli_A_S20_Slot1-20_1_8669',\n",
      "       'Clinspect_E_coli_A_S23_Slot1-8_1_8642',\n",
      "       'Clinspect_E_coli_A_S29_Slot1-19_1_8667',\n",
      "       'Clinspect_E_coli_A_S34_Slot1-18_1_8666',\n",
      "       'Clinspect_E_coli_A_S39_Slot1-26_1_8675',\n",
      "       'Clinspect_E_coli_A_S42_Slot1-21_1_8670',\n",
      "       'Clinspect_E_coli_A_S47_Slot1-14_1_8648',\n",
      "       'Clinspect_E_coli_B_QC3_Slot1-23_1_8672',\n",
      "       'Clinspect_E_coli_B_QC4_Slot1-10_1_8644',\n",
      "       'Clinspect_E_coli_B_S53_Slot1-22_1_8671',\n",
      "       'Clinspect_E_coli_B_S58_Slot1-5_1_8639',\n",
      "       'Clinspect_E_coli_B_S62_Slot1-16_1_8664',\n",
      "       'Clinspect_E_coli_B_S66_Slot1-13_1_8647',\n",
      "       'Clinspect_E_coli_B_S78_Slot1-17_1_8665',\n",
      "       'Clinspect_E_coli_B_S85_Slot1-12_1_8646',\n",
      "       'Clinspect_E_coli_B_S86_Slot1-4_1_8638',\n",
      "       'Clinspect_E_coli_B_S91_Slot1-25_1_8674',\n",
      "       'Clinspect_E_coli_B_S98_Slot1-9_1_8643'],\n",
      "      dtype='object')\n",
      "Checking batch:  lab_B\n",
      "self.design has these batch values: ['lab_A' 'lab_B']\n",
      "Got these batch indices: ['Clinspect_E_coli_A_QC1_Slot1-6_1_8640', 'Clinspect_E_coli_A_QC2_Slot1-15_1_8649', 'Clinspect_E_coli_A_S5_Slot1-7_1_8641', 'Clinspect_E_coli_A_S7_Slot1-24_1_8673', 'Clinspect_E_coli_A_S14_Slot1-11_1_8645', 'Clinspect_E_coli_A_S20_Slot1-20_1_8669', 'Clinspect_E_coli_A_S23_Slot1-8_1_8642', 'Clinspect_E_coli_A_S29_Slot1-19_1_8667', 'Clinspect_E_coli_A_S34_Slot1-18_1_8666', 'Clinspect_E_coli_A_S39_Slot1-26_1_8675', 'Clinspect_E_coli_A_S42_Slot1-21_1_8670', 'Clinspect_E_coli_A_S47_Slot1-14_1_8648', 'Clinspect_E_coli_B_QC3_Slot1-23_1_8672', 'Clinspect_E_coli_B_QC4_Slot1-10_1_8644', 'Clinspect_E_coli_B_S53_Slot1-22_1_8671', 'Clinspect_E_coli_B_S58_Slot1-5_1_8639', 'Clinspect_E_coli_B_S62_Slot1-16_1_8664', 'Clinspect_E_coli_B_S66_Slot1-13_1_8647', 'Clinspect_E_coli_B_S78_Slot1-17_1_8665', 'Clinspect_E_coli_B_S85_Slot1-12_1_8646', 'Clinspect_E_coli_B_S86_Slot1-4_1_8638', 'Clinspect_E_coli_B_S91_Slot1-25_1_8674', 'Clinspect_E_coli_B_S98_Slot1-9_1_8643']\n",
      "Trying to access data ((2910, 47))\n",
      "Data has these cols: Index(['Ref8537_QC1_20230414_1', 'Ref8537_QC2_20230414_1',\n",
      "       'Ref8537_QC3_20230414_1', 'Ref8537_QC4_20230414_1',\n",
      "       'Ref8537_S4_20230414', 'Ref8537_S8_20230414', 'Ref8537_S11_20230414',\n",
      "       'Ref8537_S18_20230414', 'Ref8537_S21_20230414', 'Ref8537_S26_20230414',\n",
      "       'Ref8537_S35_20230414', 'Ref8537_S37_20230414', 'Ref8537_S45_20230414',\n",
      "       'Ref8537_S49_20230414', 'Ref8537_S52_20230414', 'Ref8537_S60_20230414',\n",
      "       'Ref8537_S65_20230414', 'Ref8537_S69_20230414', 'Ref8537_S75_20230414',\n",
      "       'Ref8537_S79_20230414', 'Ref8537_S82_20230414', 'Ref8537_S88_20230414',\n",
      "       'Ref8537_S94_20230414', 'Ref8537_S100_20230414',\n",
      "       'Clinspect_E_coli_A_QC1_Slot1-6_1_8640',\n",
      "       'Clinspect_E_coli_A_QC2_Slot1-15_1_8649',\n",
      "       'Clinspect_E_coli_A_S5_Slot1-7_1_8641',\n",
      "       'Clinspect_E_coli_A_S7_Slot1-24_1_8673',\n",
      "       'Clinspect_E_coli_A_S14_Slot1-11_1_8645',\n",
      "       'Clinspect_E_coli_A_S20_Slot1-20_1_8669',\n",
      "       'Clinspect_E_coli_A_S23_Slot1-8_1_8642',\n",
      "       'Clinspect_E_coli_A_S29_Slot1-19_1_8667',\n",
      "       'Clinspect_E_coli_A_S34_Slot1-18_1_8666',\n",
      "       'Clinspect_E_coli_A_S39_Slot1-26_1_8675',\n",
      "       'Clinspect_E_coli_A_S42_Slot1-21_1_8670',\n",
      "       'Clinspect_E_coli_A_S47_Slot1-14_1_8648',\n",
      "       'Clinspect_E_coli_B_QC3_Slot1-23_1_8672',\n",
      "       'Clinspect_E_coli_B_QC4_Slot1-10_1_8644',\n",
      "       'Clinspect_E_coli_B_S53_Slot1-22_1_8671',\n",
      "       'Clinspect_E_coli_B_S58_Slot1-5_1_8639',\n",
      "       'Clinspect_E_coli_B_S62_Slot1-16_1_8664',\n",
      "       'Clinspect_E_coli_B_S66_Slot1-13_1_8647',\n",
      "       'Clinspect_E_coli_B_S78_Slot1-17_1_8665',\n",
      "       'Clinspect_E_coli_B_S85_Slot1-12_1_8646',\n",
      "       'Clinspect_E_coli_B_S86_Slot1-4_1_8638',\n",
      "       'Clinspect_E_coli_B_S91_Slot1-25_1_8674',\n",
      "       'Clinspect_E_coli_B_S98_Slot1-9_1_8643'],\n",
      "      dtype='object')\n",
      "Got the following config:\n",
      "{'flimmaBatchCorrection': {'batch_col': 'batch', 'covariates': ['Pyr'], 'data_filename': 'intensities_log_UNION.tsv', 'design_filename': 'design.tsv', 'design_separator': '\\t', 'expression_file_flag': True, 'index_col': 'rowname', 'min_samples': 0, 'normalizationMethod': None, 'position': 1, 'separator': '\\t', 'smpc': False}}\n",
      "Opening dataset ../evaluation_data/proteomics_multibatch/before/center2/intensities_log_UNION.tsv\n",
      "Shape of rawdata(expr_file): (2900, 47)\n",
      "finished loading data, shape of data: (2900, 47), num_features: 2900, num_samples: 47\n",
      "Checking batch:  lab_C\n",
      "self.design has these batch values: ['lab_C' 'lab_D']\n",
      "Got these batch indices: ['BBM_673_P283_01_VEB_001_R4', 'BBM_673_P283_01_VEB_002_R3', 'BBM_673_P283_01_VEB_003_R3', 'BBM_673_P283_01_VEB_004_R3', 'BBM_673_P283_01_VEB_005_R3', 'BBM_673_P283_01_VEB_006_R2', 'BBM_673_P283_01_VEB_007_R2', 'BBM_673_P283_01_VEB_009_R2', 'BBM_673_P283_01_VEB_010_R2', 'BBM_673_P283_01_VEB_011_R2', 'BBM_673_P283_01_VEB_012_R2', 'BBM_673_P283_01_VEB_013_R2', 'BBM_673_P283_01_VEB_014_R2', 'BBM_673_P283_01_VEB_015_R2', 'BBM_673_P283_01_VEB_016_R2', 'BBM_673_P283_01_VEB_017_R2', 'BBM_673_P283_01_VEB_018_R2', 'BBM_673_P283_01_VEB_019_R2', 'BBM_673_P283_01_VEB_020_R2', 'BBM_673_P283_01_VEB_021_R2', 'BBM_673_P283_01_VEB_022_R3', 'BBM_673_P283_01_VEB_023_R3', 'BBM_673_P283_01_VEB_024_R3']\n",
      "Trying to access data ((2900, 47))\n",
      "Data has these cols: Index(['BBM_673_P283_01_VEB_001_R4', 'BBM_673_P283_01_VEB_002_R3',\n",
      "       'BBM_673_P283_01_VEB_003_R3', 'BBM_673_P283_01_VEB_004_R3',\n",
      "       'BBM_673_P283_01_VEB_005_R3', 'BBM_673_P283_01_VEB_006_R2',\n",
      "       'BBM_673_P283_01_VEB_007_R2', 'BBM_673_P283_01_VEB_009_R2',\n",
      "       'BBM_673_P283_01_VEB_010_R2', 'BBM_673_P283_01_VEB_011_R2',\n",
      "       'BBM_673_P283_01_VEB_012_R2', 'BBM_673_P283_01_VEB_013_R2',\n",
      "       'BBM_673_P283_01_VEB_014_R2', 'BBM_673_P283_01_VEB_015_R2',\n",
      "       'BBM_673_P283_01_VEB_016_R2', 'BBM_673_P283_01_VEB_017_R2',\n",
      "       'BBM_673_P283_01_VEB_018_R2', 'BBM_673_P283_01_VEB_019_R2',\n",
      "       'BBM_673_P283_01_VEB_020_R2', 'BBM_673_P283_01_VEB_021_R2',\n",
      "       'BBM_673_P283_01_VEB_022_R3', 'BBM_673_P283_01_VEB_023_R3',\n",
      "       'BBM_673_P283_01_VEB_024_R3', '20230516_EXPL1_LiSc_SA_S610_RT_25_s2',\n",
      "       '20230516_EXPL1_LiSc_SA_S610_RT_26_s9',\n",
      "       '20230516_EXPL1_LiSc_SA_S610_RT_27_s12',\n",
      "       '20230516_EXPL1_LiSc_SA_S610_RT_28_s17',\n",
      "       '20230516_EXPL1_LiSc_SA_S610_RT_29_s24',\n",
      "       '20230516_EXPL1_LiSc_SA_S610_RT_30_s30',\n",
      "       '20230516_EXPL1_LiSc_SA_S610_RT_31_s33',\n",
      "       '20230516_EXPL1_LiSc_SA_S610_RT_32_s38',\n",
      "       '20230516_EXPL1_LiSc_SA_S610_RT_33_s44',\n",
      "       '20230516_EXPL1_LiSc_SA_S610_RT_34_s48',\n",
      "       '20230516_EXPL1_LiSc_SA_S610_RT_35_s51',\n",
      "       '20230516_EXPL1_LiSc_SA_S610_RT_36_s56',\n",
      "       '20230516_EXPL1_LiSc_SA_S610_RT_37_s63',\n",
      "       '20230516_EXPL1_LiSc_SA_S610_RT_38_s68',\n",
      "       '20230516_EXPL1_LiSc_SA_S610_RT_39_s72',\n",
      "       '20230516_EXPL1_LiSc_SA_S610_RT_40_s80',\n",
      "       '20230516_EXPL1_LiSc_SA_S610_RT_41_s81',\n",
      "       '20230516_EXPL1_LiSc_SA_S610_RT_42_s89',\n",
      "       '20230516_EXPL1_LiSc_SA_S610_RT_43_s92',\n",
      "       '20230516_EXPL1_LiSc_SA_S610_RT_44_s96',\n",
      "       '20230516_EXPL1_LiSc_SA_S610_RT_45_QC1_LabD',\n",
      "       '20230516_EXPL1_LiSc_SA_S610_RT_46_QC2_LabD',\n",
      "       '20230516_EXPL1_LiSc_SA_S610_RT_47_QC3_LabD',\n",
      "       '20230516_EXPL1_LiSc_SA_S610_RT_48_QC4_LabD'],\n",
      "      dtype='object')\n",
      "Checking batch:  lab_D\n",
      "self.design has these batch values: ['lab_C' 'lab_D']\n",
      "Got these batch indices: ['20230516_EXPL1_LiSc_SA_S610_RT_25_s2', '20230516_EXPL1_LiSc_SA_S610_RT_26_s9', '20230516_EXPL1_LiSc_SA_S610_RT_27_s12', '20230516_EXPL1_LiSc_SA_S610_RT_28_s17', '20230516_EXPL1_LiSc_SA_S610_RT_29_s24', '20230516_EXPL1_LiSc_SA_S610_RT_30_s30', '20230516_EXPL1_LiSc_SA_S610_RT_31_s33', '20230516_EXPL1_LiSc_SA_S610_RT_32_s38', '20230516_EXPL1_LiSc_SA_S610_RT_33_s44', '20230516_EXPL1_LiSc_SA_S610_RT_34_s48', '20230516_EXPL1_LiSc_SA_S610_RT_35_s51', '20230516_EXPL1_LiSc_SA_S610_RT_36_s56', '20230516_EXPL1_LiSc_SA_S610_RT_37_s63', '20230516_EXPL1_LiSc_SA_S610_RT_38_s68', '20230516_EXPL1_LiSc_SA_S610_RT_39_s72', '20230516_EXPL1_LiSc_SA_S610_RT_40_s80', '20230516_EXPL1_LiSc_SA_S610_RT_41_s81', '20230516_EXPL1_LiSc_SA_S610_RT_42_s89', '20230516_EXPL1_LiSc_SA_S610_RT_43_s92', '20230516_EXPL1_LiSc_SA_S610_RT_44_s96', '20230516_EXPL1_LiSc_SA_S610_RT_45_QC1_LabD', '20230516_EXPL1_LiSc_SA_S610_RT_46_QC2_LabD', '20230516_EXPL1_LiSc_SA_S610_RT_47_QC3_LabD', '20230516_EXPL1_LiSc_SA_S610_RT_48_QC4_LabD']\n",
      "Trying to access data ((2900, 47))\n",
      "Data has these cols: Index(['BBM_673_P283_01_VEB_001_R4', 'BBM_673_P283_01_VEB_002_R3',\n",
      "       'BBM_673_P283_01_VEB_003_R3', 'BBM_673_P283_01_VEB_004_R3',\n",
      "       'BBM_673_P283_01_VEB_005_R3', 'BBM_673_P283_01_VEB_006_R2',\n",
      "       'BBM_673_P283_01_VEB_007_R2', 'BBM_673_P283_01_VEB_009_R2',\n",
      "       'BBM_673_P283_01_VEB_010_R2', 'BBM_673_P283_01_VEB_011_R2',\n",
      "       'BBM_673_P283_01_VEB_012_R2', 'BBM_673_P283_01_VEB_013_R2',\n",
      "       'BBM_673_P283_01_VEB_014_R2', 'BBM_673_P283_01_VEB_015_R2',\n",
      "       'BBM_673_P283_01_VEB_016_R2', 'BBM_673_P283_01_VEB_017_R2',\n",
      "       'BBM_673_P283_01_VEB_018_R2', 'BBM_673_P283_01_VEB_019_R2',\n",
      "       'BBM_673_P283_01_VEB_020_R2', 'BBM_673_P283_01_VEB_021_R2',\n",
      "       'BBM_673_P283_01_VEB_022_R3', 'BBM_673_P283_01_VEB_023_R3',\n",
      "       'BBM_673_P283_01_VEB_024_R3', '20230516_EXPL1_LiSc_SA_S610_RT_25_s2',\n",
      "       '20230516_EXPL1_LiSc_SA_S610_RT_26_s9',\n",
      "       '20230516_EXPL1_LiSc_SA_S610_RT_27_s12',\n",
      "       '20230516_EXPL1_LiSc_SA_S610_RT_28_s17',\n",
      "       '20230516_EXPL1_LiSc_SA_S610_RT_29_s24',\n",
      "       '20230516_EXPL1_LiSc_SA_S610_RT_30_s30',\n",
      "       '20230516_EXPL1_LiSc_SA_S610_RT_31_s33',\n",
      "       '20230516_EXPL1_LiSc_SA_S610_RT_32_s38',\n",
      "       '20230516_EXPL1_LiSc_SA_S610_RT_33_s44',\n",
      "       '20230516_EXPL1_LiSc_SA_S610_RT_34_s48',\n",
      "       '20230516_EXPL1_LiSc_SA_S610_RT_35_s51',\n",
      "       '20230516_EXPL1_LiSc_SA_S610_RT_36_s56',\n",
      "       '20230516_EXPL1_LiSc_SA_S610_RT_37_s63',\n",
      "       '20230516_EXPL1_LiSc_SA_S610_RT_38_s68',\n",
      "       '20230516_EXPL1_LiSc_SA_S610_RT_39_s72',\n",
      "       '20230516_EXPL1_LiSc_SA_S610_RT_40_s80',\n",
      "       '20230516_EXPL1_LiSc_SA_S610_RT_41_s81',\n",
      "       '20230516_EXPL1_LiSc_SA_S610_RT_42_s89',\n",
      "       '20230516_EXPL1_LiSc_SA_S610_RT_43_s92',\n",
      "       '20230516_EXPL1_LiSc_SA_S610_RT_44_s96',\n",
      "       '20230516_EXPL1_LiSc_SA_S610_RT_45_QC1_LabD',\n",
      "       '20230516_EXPL1_LiSc_SA_S610_RT_46_QC2_LabD',\n",
      "       '20230516_EXPL1_LiSc_SA_S610_RT_47_QC3_LabD',\n",
      "       '20230516_EXPL1_LiSc_SA_S610_RT_48_QC4_LabD'],\n",
      "      dtype='object')\n",
      "Got the following config:\n",
      "{'flimmaBatchCorrection': {'batch_col': 'batch', 'covariates': ['Pyr'], 'data_filename': 'intensities_log_UNION.tsv', 'design_filename': 'design.tsv', 'design_separator': '\\t', 'expression_file_flag': True, 'index_col': 'rowname', 'min_samples': 0, 'normalizationMethod': None, 'position': 2, 'separator': '\\t', 'smpc': False}}\n",
      "Opening dataset ../evaluation_data/proteomics_multibatch/before/center3/intensities_log_UNION.tsv\n",
      "Shape of rawdata(expr_file): (2363, 24)\n",
      "finished loading data, shape of data: (2363, 24), num_features: 2363, num_samples: 24\n"
     ]
    }
   ],
   "source": [
    "### SIMULATION: all: initial ###\n",
    "### Initial reading of the input folder\n",
    "\n",
    "send_feature_variable_batch_info = list()\n",
    "for clientWrapper in clientWrappers:\n",
    "    # define the client class\n",
    "    cohort_name = clientWrapper.id\n",
    "    client = Client()\n",
    "    client.config_based_init(clientname = cohort_name,\n",
    "                             input_folder = clientWrapper.input_folder,\n",
    "                             use_hashing = False)\n",
    "    clientWrapper.client_class = client\n",
    "    assert isinstance(client.hash2feature, dict)\n",
    "    assert isinstance(client.hash2variable, dict)\n",
    "    batch_feature_presence_info: Dict[str, List[str]] = client.get_batch_feature_presence_info()\n",
    "    send_feature_variable_batch_info.append((cohort_name,\n",
    "                                        list(client.hash2variable.keys()),\n",
    "                                        client.position,\n",
    "                                        batch_feature_presence_info))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO: All cohorts: [['center1|lab_A', 'center1|lab_B'], ['center2|lab_C', 'center2|lab_D'], ['center3|lab_E']]\n",
      "Using given specific client order: ['center1', 'center2', 'center3']\n",
      "INFO: Cohorts order: ['center1|lab_A', 'center1|lab_B', 'center2|lab_C', 'center2|lab_D', 'center3|lab_E']\n",
      "Got this final cohort order: ['center1|lab_A', 'center1|lab_B', 'center2|lab_C', 'center2|lab_D', 'center3|lab_E']\n"
     ]
    }
   ],
   "source": [
    "### SIMULATION: Coordinator: global_feature_selection ###\n",
    "### Aggregate the features and variables\n",
    "\n",
    "# obtain and safe common genes and indices of design matrix\n",
    "# wait for each client to send the list of genes they have\n",
    "# also memo the feature presence matrix and feature_to_cohorts\n",
    "\n",
    "broadcast_features_variables = tuple()\n",
    "for clientWrapper in clientWrappers:\n",
    "    if clientWrapper.is_coordinator:\n",
    "\n",
    "        time_tracker[\"Coordinator\"] = time.time()\n",
    "\n",
    "        global_feature_names, global_variables, feature_presence_matrix, cohorts_order = \\\n",
    "            select_common_features_variables(\n",
    "                feature_variable_batch_info=send_feature_variable_batch_info,\n",
    "                min_clients=1,\n",
    "                default_order=cohort_names\n",
    "            )\n",
    "\n",
    "        # memo the feature presence matrix and feature_to_cohorts\n",
    "        broadcast_features_variables = global_feature_names, global_variables, cohorts_order\n",
    "        print(f\"Got this final cohort order: {cohorts_order}\")\n",
    "        end_time = time.time()\n",
    "        time_tracker[\"Coordinator\"] = end_time - time_tracker[\"Coordinator\"]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Client center1: Inputs validated.\n",
      "Number of features available in this client: 2910\n",
      "Number of features given globally: 3034\n",
      "Number of features only available on this client: 0\n",
      "Number of features available in other clients but not this client: 124\n",
      "Adding 124 extra global features\n",
      "Design matrix created:\n",
      "                                        intercept  Pyr  center1|lab_A  \\\n",
      "file                                                                    \n",
      "Ref8537_QC1_20230414_1                        1.0    1              1   \n",
      "Ref8537_QC2_20230414_1                        1.0    1              1   \n",
      "Ref8537_QC3_20230414_1                        1.0    0              1   \n",
      "Ref8537_QC4_20230414_1                        1.0    0              1   \n",
      "Ref8537_S4_20230414                           1.0    1              1   \n",
      "Ref8537_S8_20230414                           1.0    1              1   \n",
      "Ref8537_S11_20230414                          1.0    1              1   \n",
      "Ref8537_S18_20230414                          1.0    1              1   \n",
      "Ref8537_S21_20230414                          1.0    1              1   \n",
      "Ref8537_S26_20230414                          1.0    1              1   \n",
      "Ref8537_S35_20230414                          1.0    1              1   \n",
      "Ref8537_S37_20230414                          1.0    1              1   \n",
      "Ref8537_S45_20230414                          1.0    1              1   \n",
      "Ref8537_S49_20230414                          1.0    1              1   \n",
      "Ref8537_S52_20230414                          1.0    0              1   \n",
      "Ref8537_S60_20230414                          1.0    0              1   \n",
      "Ref8537_S65_20230414                          1.0    0              1   \n",
      "Ref8537_S69_20230414                          1.0    0              1   \n",
      "Ref8537_S75_20230414                          1.0    0              1   \n",
      "Ref8537_S79_20230414                          1.0    0              1   \n",
      "Ref8537_S82_20230414                          1.0    0              1   \n",
      "Ref8537_S88_20230414                          1.0    0              1   \n",
      "Ref8537_S94_20230414                          1.0    0              1   \n",
      "Ref8537_S100_20230414                         1.0    0              1   \n",
      "Clinspect_E_coli_A_QC1_Slot1-6_1_8640         1.0    1              0   \n",
      "Clinspect_E_coli_A_QC2_Slot1-15_1_8649        1.0    1              0   \n",
      "Clinspect_E_coli_A_S5_Slot1-7_1_8641          1.0    1              0   \n",
      "Clinspect_E_coli_A_S7_Slot1-24_1_8673         1.0    1              0   \n",
      "Clinspect_E_coli_A_S14_Slot1-11_1_8645        1.0    1              0   \n",
      "Clinspect_E_coli_A_S20_Slot1-20_1_8669        1.0    1              0   \n",
      "Clinspect_E_coli_A_S23_Slot1-8_1_8642         1.0    1              0   \n",
      "Clinspect_E_coli_A_S29_Slot1-19_1_8667        1.0    1              0   \n",
      "Clinspect_E_coli_A_S34_Slot1-18_1_8666        1.0    1              0   \n",
      "Clinspect_E_coli_A_S39_Slot1-26_1_8675        1.0    1              0   \n",
      "Clinspect_E_coli_A_S42_Slot1-21_1_8670        1.0    1              0   \n",
      "Clinspect_E_coli_A_S47_Slot1-14_1_8648        1.0    1              0   \n",
      "Clinspect_E_coli_B_QC3_Slot1-23_1_8672        1.0    0              0   \n",
      "Clinspect_E_coli_B_QC4_Slot1-10_1_8644        1.0    0              0   \n",
      "Clinspect_E_coli_B_S53_Slot1-22_1_8671        1.0    0              0   \n",
      "Clinspect_E_coli_B_S58_Slot1-5_1_8639         1.0    0              0   \n",
      "Clinspect_E_coli_B_S62_Slot1-16_1_8664        1.0    0              0   \n",
      "Clinspect_E_coli_B_S66_Slot1-13_1_8647        1.0    0              0   \n",
      "Clinspect_E_coli_B_S78_Slot1-17_1_8665        1.0    0              0   \n",
      "Clinspect_E_coli_B_S85_Slot1-12_1_8646        1.0    0              0   \n",
      "Clinspect_E_coli_B_S86_Slot1-4_1_8638         1.0    0              0   \n",
      "Clinspect_E_coli_B_S91_Slot1-25_1_8674        1.0    0              0   \n",
      "Clinspect_E_coli_B_S98_Slot1-9_1_8643         1.0    0              0   \n",
      "\n",
      "                                        center1|lab_B  center2|lab_C  \\\n",
      "file                                                                   \n",
      "Ref8537_QC1_20230414_1                              0              0   \n",
      "Ref8537_QC2_20230414_1                              0              0   \n",
      "Ref8537_QC3_20230414_1                              0              0   \n",
      "Ref8537_QC4_20230414_1                              0              0   \n",
      "Ref8537_S4_20230414                                 0              0   \n",
      "Ref8537_S8_20230414                                 0              0   \n",
      "Ref8537_S11_20230414                                0              0   \n",
      "Ref8537_S18_20230414                                0              0   \n",
      "Ref8537_S21_20230414                                0              0   \n",
      "Ref8537_S26_20230414                                0              0   \n",
      "Ref8537_S35_20230414                                0              0   \n",
      "Ref8537_S37_20230414                                0              0   \n",
      "Ref8537_S45_20230414                                0              0   \n",
      "Ref8537_S49_20230414                                0              0   \n",
      "Ref8537_S52_20230414                                0              0   \n",
      "Ref8537_S60_20230414                                0              0   \n",
      "Ref8537_S65_20230414                                0              0   \n",
      "Ref8537_S69_20230414                                0              0   \n",
      "Ref8537_S75_20230414                                0              0   \n",
      "Ref8537_S79_20230414                                0              0   \n",
      "Ref8537_S82_20230414                                0              0   \n",
      "Ref8537_S88_20230414                                0              0   \n",
      "Ref8537_S94_20230414                                0              0   \n",
      "Ref8537_S100_20230414                               0              0   \n",
      "Clinspect_E_coli_A_QC1_Slot1-6_1_8640               1              0   \n",
      "Clinspect_E_coli_A_QC2_Slot1-15_1_8649              1              0   \n",
      "Clinspect_E_coli_A_S5_Slot1-7_1_8641                1              0   \n",
      "Clinspect_E_coli_A_S7_Slot1-24_1_8673               1              0   \n",
      "Clinspect_E_coli_A_S14_Slot1-11_1_8645              1              0   \n",
      "Clinspect_E_coli_A_S20_Slot1-20_1_8669              1              0   \n",
      "Clinspect_E_coli_A_S23_Slot1-8_1_8642               1              0   \n",
      "Clinspect_E_coli_A_S29_Slot1-19_1_8667              1              0   \n",
      "Clinspect_E_coli_A_S34_Slot1-18_1_8666              1              0   \n",
      "Clinspect_E_coli_A_S39_Slot1-26_1_8675              1              0   \n",
      "Clinspect_E_coli_A_S42_Slot1-21_1_8670              1              0   \n",
      "Clinspect_E_coli_A_S47_Slot1-14_1_8648              1              0   \n",
      "Clinspect_E_coli_B_QC3_Slot1-23_1_8672              1              0   \n",
      "Clinspect_E_coli_B_QC4_Slot1-10_1_8644              1              0   \n",
      "Clinspect_E_coli_B_S53_Slot1-22_1_8671              1              0   \n",
      "Clinspect_E_coli_B_S58_Slot1-5_1_8639               1              0   \n",
      "Clinspect_E_coli_B_S62_Slot1-16_1_8664              1              0   \n",
      "Clinspect_E_coli_B_S66_Slot1-13_1_8647              1              0   \n",
      "Clinspect_E_coli_B_S78_Slot1-17_1_8665              1              0   \n",
      "Clinspect_E_coli_B_S85_Slot1-12_1_8646              1              0   \n",
      "Clinspect_E_coli_B_S86_Slot1-4_1_8638               1              0   \n",
      "Clinspect_E_coli_B_S91_Slot1-25_1_8674              1              0   \n",
      "Clinspect_E_coli_B_S98_Slot1-9_1_8643               1              0   \n",
      "\n",
      "                                        center2|lab_D  \n",
      "file                                                   \n",
      "Ref8537_QC1_20230414_1                              0  \n",
      "Ref8537_QC2_20230414_1                              0  \n",
      "Ref8537_QC3_20230414_1                              0  \n",
      "Ref8537_QC4_20230414_1                              0  \n",
      "Ref8537_S4_20230414                                 0  \n",
      "Ref8537_S8_20230414                                 0  \n",
      "Ref8537_S11_20230414                                0  \n",
      "Ref8537_S18_20230414                                0  \n",
      "Ref8537_S21_20230414                                0  \n",
      "Ref8537_S26_20230414                                0  \n",
      "Ref8537_S35_20230414                                0  \n",
      "Ref8537_S37_20230414                                0  \n",
      "Ref8537_S45_20230414                                0  \n",
      "Ref8537_S49_20230414                                0  \n",
      "Ref8537_S52_20230414                                0  \n",
      "Ref8537_S60_20230414                                0  \n",
      "Ref8537_S65_20230414                                0  \n",
      "Ref8537_S69_20230414                                0  \n",
      "Ref8537_S75_20230414                                0  \n",
      "Ref8537_S79_20230414                                0  \n",
      "Ref8537_S82_20230414                                0  \n",
      "Ref8537_S88_20230414                                0  \n",
      "Ref8537_S94_20230414                                0  \n",
      "Ref8537_S100_20230414                               0  \n",
      "Clinspect_E_coli_A_QC1_Slot1-6_1_8640               0  \n",
      "Clinspect_E_coli_A_QC2_Slot1-15_1_8649              0  \n",
      "Clinspect_E_coli_A_S5_Slot1-7_1_8641                0  \n",
      "Clinspect_E_coli_A_S7_Slot1-24_1_8673               0  \n",
      "Clinspect_E_coli_A_S14_Slot1-11_1_8645              0  \n",
      "Clinspect_E_coli_A_S20_Slot1-20_1_8669              0  \n",
      "Clinspect_E_coli_A_S23_Slot1-8_1_8642               0  \n",
      "Clinspect_E_coli_A_S29_Slot1-19_1_8667              0  \n",
      "Clinspect_E_coli_A_S34_Slot1-18_1_8666              0  \n",
      "Clinspect_E_coli_A_S39_Slot1-26_1_8675              0  \n",
      "Clinspect_E_coli_A_S42_Slot1-21_1_8670              0  \n",
      "Clinspect_E_coli_A_S47_Slot1-14_1_8648              0  \n",
      "Clinspect_E_coli_B_QC3_Slot1-23_1_8672              0  \n",
      "Clinspect_E_coli_B_QC4_Slot1-10_1_8644              0  \n",
      "Clinspect_E_coli_B_S53_Slot1-22_1_8671              0  \n",
      "Clinspect_E_coli_B_S58_Slot1-5_1_8639               0  \n",
      "Clinspect_E_coli_B_S62_Slot1-16_1_8664              0  \n",
      "Clinspect_E_coli_B_S66_Slot1-13_1_8647              0  \n",
      "Clinspect_E_coli_B_S78_Slot1-17_1_8665              0  \n",
      "Clinspect_E_coli_B_S85_Slot1-12_1_8646              0  \n",
      "Clinspect_E_coli_B_S86_Slot1-4_1_8638               0  \n",
      "Clinspect_E_coli_B_S91_Slot1-25_1_8674              0  \n",
      "Clinspect_E_coli_B_S98_Slot1-9_1_8643               0  \n",
      "Design matrix shape: (47, 6)\n",
      "Design matrix with summed columns: intercept        47.0\n",
      "Pyr              24.0\n",
      "center1|lab_A    24.0\n",
      "center1|lab_B    23.0\n",
      "center2|lab_C     0.0\n",
      "center2|lab_D     0.0\n",
      "dtype: float64\n",
      "Client center2: Inputs validated.\n",
      "Number of features available in this client: 2900\n",
      "Number of features given globally: 3034\n",
      "Number of features only available on this client: 0\n",
      "Number of features available in other clients but not this client: 134\n",
      "Adding 134 extra global features\n",
      "Design matrix created:\n",
      "                                            intercept  Pyr  center1|lab_A  \\\n",
      "file                                                                        \n",
      "BBM_673_P283_01_VEB_001_R4                        1.0    1              0   \n",
      "BBM_673_P283_01_VEB_002_R3                        1.0    1              0   \n",
      "BBM_673_P283_01_VEB_003_R3                        1.0    1              0   \n",
      "BBM_673_P283_01_VEB_004_R3                        1.0    1              0   \n",
      "BBM_673_P283_01_VEB_005_R3                        1.0    1              0   \n",
      "BBM_673_P283_01_VEB_006_R2                        1.0    1              0   \n",
      "BBM_673_P283_01_VEB_007_R2                        1.0    1              0   \n",
      "BBM_673_P283_01_VEB_009_R2                        1.0    1              0   \n",
      "BBM_673_P283_01_VEB_010_R2                        1.0    1              0   \n",
      "BBM_673_P283_01_VEB_011_R2                        1.0    1              0   \n",
      "BBM_673_P283_01_VEB_012_R2                        1.0    1              0   \n",
      "BBM_673_P283_01_VEB_013_R2                        1.0    0              0   \n",
      "BBM_673_P283_01_VEB_014_R2                        1.0    0              0   \n",
      "BBM_673_P283_01_VEB_015_R2                        1.0    0              0   \n",
      "BBM_673_P283_01_VEB_016_R2                        1.0    0              0   \n",
      "BBM_673_P283_01_VEB_017_R2                        1.0    0              0   \n",
      "BBM_673_P283_01_VEB_018_R2                        1.0    0              0   \n",
      "BBM_673_P283_01_VEB_019_R2                        1.0    0              0   \n",
      "BBM_673_P283_01_VEB_020_R2                        1.0    0              0   \n",
      "BBM_673_P283_01_VEB_021_R2                        1.0    0              0   \n",
      "BBM_673_P283_01_VEB_022_R3                        1.0    0              0   \n",
      "BBM_673_P283_01_VEB_023_R3                        1.0    0              0   \n",
      "BBM_673_P283_01_VEB_024_R3                        1.0    0              0   \n",
      "20230516_EXPL1_LiSc_SA_S610_RT_25_s2              1.0    1              0   \n",
      "20230516_EXPL1_LiSc_SA_S610_RT_26_s9              1.0    1              0   \n",
      "20230516_EXPL1_LiSc_SA_S610_RT_27_s12             1.0    1              0   \n",
      "20230516_EXPL1_LiSc_SA_S610_RT_28_s17             1.0    1              0   \n",
      "20230516_EXPL1_LiSc_SA_S610_RT_29_s24             1.0    1              0   \n",
      "20230516_EXPL1_LiSc_SA_S610_RT_30_s30             1.0    1              0   \n",
      "20230516_EXPL1_LiSc_SA_S610_RT_31_s33             1.0    1              0   \n",
      "20230516_EXPL1_LiSc_SA_S610_RT_32_s38             1.0    1              0   \n",
      "20230516_EXPL1_LiSc_SA_S610_RT_33_s44             1.0    1              0   \n",
      "20230516_EXPL1_LiSc_SA_S610_RT_34_s48             1.0    1              0   \n",
      "20230516_EXPL1_LiSc_SA_S610_RT_35_s51             1.0    0              0   \n",
      "20230516_EXPL1_LiSc_SA_S610_RT_36_s56             1.0    0              0   \n",
      "20230516_EXPL1_LiSc_SA_S610_RT_37_s63             1.0    0              0   \n",
      "20230516_EXPL1_LiSc_SA_S610_RT_38_s68             1.0    0              0   \n",
      "20230516_EXPL1_LiSc_SA_S610_RT_39_s72             1.0    0              0   \n",
      "20230516_EXPL1_LiSc_SA_S610_RT_40_s80             1.0    0              0   \n",
      "20230516_EXPL1_LiSc_SA_S610_RT_41_s81             1.0    0              0   \n",
      "20230516_EXPL1_LiSc_SA_S610_RT_42_s89             1.0    0              0   \n",
      "20230516_EXPL1_LiSc_SA_S610_RT_43_s92             1.0    0              0   \n",
      "20230516_EXPL1_LiSc_SA_S610_RT_44_s96             1.0    0              0   \n",
      "20230516_EXPL1_LiSc_SA_S610_RT_45_QC1_LabD        1.0    1              0   \n",
      "20230516_EXPL1_LiSc_SA_S610_RT_46_QC2_LabD        1.0    1              0   \n",
      "20230516_EXPL1_LiSc_SA_S610_RT_47_QC3_LabD        1.0    0              0   \n",
      "20230516_EXPL1_LiSc_SA_S610_RT_48_QC4_LabD        1.0    0              0   \n",
      "\n",
      "                                            center1|lab_B  center2|lab_C  \\\n",
      "file                                                                       \n",
      "BBM_673_P283_01_VEB_001_R4                              0              1   \n",
      "BBM_673_P283_01_VEB_002_R3                              0              1   \n",
      "BBM_673_P283_01_VEB_003_R3                              0              1   \n",
      "BBM_673_P283_01_VEB_004_R3                              0              1   \n",
      "BBM_673_P283_01_VEB_005_R3                              0              1   \n",
      "BBM_673_P283_01_VEB_006_R2                              0              1   \n",
      "BBM_673_P283_01_VEB_007_R2                              0              1   \n",
      "BBM_673_P283_01_VEB_009_R2                              0              1   \n",
      "BBM_673_P283_01_VEB_010_R2                              0              1   \n",
      "BBM_673_P283_01_VEB_011_R2                              0              1   \n",
      "BBM_673_P283_01_VEB_012_R2                              0              1   \n",
      "BBM_673_P283_01_VEB_013_R2                              0              1   \n",
      "BBM_673_P283_01_VEB_014_R2                              0              1   \n",
      "BBM_673_P283_01_VEB_015_R2                              0              1   \n",
      "BBM_673_P283_01_VEB_016_R2                              0              1   \n",
      "BBM_673_P283_01_VEB_017_R2                              0              1   \n",
      "BBM_673_P283_01_VEB_018_R2                              0              1   \n",
      "BBM_673_P283_01_VEB_019_R2                              0              1   \n",
      "BBM_673_P283_01_VEB_020_R2                              0              1   \n",
      "BBM_673_P283_01_VEB_021_R2                              0              1   \n",
      "BBM_673_P283_01_VEB_022_R3                              0              1   \n",
      "BBM_673_P283_01_VEB_023_R3                              0              1   \n",
      "BBM_673_P283_01_VEB_024_R3                              0              1   \n",
      "20230516_EXPL1_LiSc_SA_S610_RT_25_s2                    0              0   \n",
      "20230516_EXPL1_LiSc_SA_S610_RT_26_s9                    0              0   \n",
      "20230516_EXPL1_LiSc_SA_S610_RT_27_s12                   0              0   \n",
      "20230516_EXPL1_LiSc_SA_S610_RT_28_s17                   0              0   \n",
      "20230516_EXPL1_LiSc_SA_S610_RT_29_s24                   0              0   \n",
      "20230516_EXPL1_LiSc_SA_S610_RT_30_s30                   0              0   \n",
      "20230516_EXPL1_LiSc_SA_S610_RT_31_s33                   0              0   \n",
      "20230516_EXPL1_LiSc_SA_S610_RT_32_s38                   0              0   \n",
      "20230516_EXPL1_LiSc_SA_S610_RT_33_s44                   0              0   \n",
      "20230516_EXPL1_LiSc_SA_S610_RT_34_s48                   0              0   \n",
      "20230516_EXPL1_LiSc_SA_S610_RT_35_s51                   0              0   \n",
      "20230516_EXPL1_LiSc_SA_S610_RT_36_s56                   0              0   \n",
      "20230516_EXPL1_LiSc_SA_S610_RT_37_s63                   0              0   \n",
      "20230516_EXPL1_LiSc_SA_S610_RT_38_s68                   0              0   \n",
      "20230516_EXPL1_LiSc_SA_S610_RT_39_s72                   0              0   \n",
      "20230516_EXPL1_LiSc_SA_S610_RT_40_s80                   0              0   \n",
      "20230516_EXPL1_LiSc_SA_S610_RT_41_s81                   0              0   \n",
      "20230516_EXPL1_LiSc_SA_S610_RT_42_s89                   0              0   \n",
      "20230516_EXPL1_LiSc_SA_S610_RT_43_s92                   0              0   \n",
      "20230516_EXPL1_LiSc_SA_S610_RT_44_s96                   0              0   \n",
      "20230516_EXPL1_LiSc_SA_S610_RT_45_QC1_LabD              0              0   \n",
      "20230516_EXPL1_LiSc_SA_S610_RT_46_QC2_LabD              0              0   \n",
      "20230516_EXPL1_LiSc_SA_S610_RT_47_QC3_LabD              0              0   \n",
      "20230516_EXPL1_LiSc_SA_S610_RT_48_QC4_LabD              0              0   \n",
      "\n",
      "                                            center2|lab_D  \n",
      "file                                                       \n",
      "BBM_673_P283_01_VEB_001_R4                              0  \n",
      "BBM_673_P283_01_VEB_002_R3                              0  \n",
      "BBM_673_P283_01_VEB_003_R3                              0  \n",
      "BBM_673_P283_01_VEB_004_R3                              0  \n",
      "BBM_673_P283_01_VEB_005_R3                              0  \n",
      "BBM_673_P283_01_VEB_006_R2                              0  \n",
      "BBM_673_P283_01_VEB_007_R2                              0  \n",
      "BBM_673_P283_01_VEB_009_R2                              0  \n",
      "BBM_673_P283_01_VEB_010_R2                              0  \n",
      "BBM_673_P283_01_VEB_011_R2                              0  \n",
      "BBM_673_P283_01_VEB_012_R2                              0  \n",
      "BBM_673_P283_01_VEB_013_R2                              0  \n",
      "BBM_673_P283_01_VEB_014_R2                              0  \n",
      "BBM_673_P283_01_VEB_015_R2                              0  \n",
      "BBM_673_P283_01_VEB_016_R2                              0  \n",
      "BBM_673_P283_01_VEB_017_R2                              0  \n",
      "BBM_673_P283_01_VEB_018_R2                              0  \n",
      "BBM_673_P283_01_VEB_019_R2                              0  \n",
      "BBM_673_P283_01_VEB_020_R2                              0  \n",
      "BBM_673_P283_01_VEB_021_R2                              0  \n",
      "BBM_673_P283_01_VEB_022_R3                              0  \n",
      "BBM_673_P283_01_VEB_023_R3                              0  \n",
      "BBM_673_P283_01_VEB_024_R3                              0  \n",
      "20230516_EXPL1_LiSc_SA_S610_RT_25_s2                    1  \n",
      "20230516_EXPL1_LiSc_SA_S610_RT_26_s9                    1  \n",
      "20230516_EXPL1_LiSc_SA_S610_RT_27_s12                   1  \n",
      "20230516_EXPL1_LiSc_SA_S610_RT_28_s17                   1  \n",
      "20230516_EXPL1_LiSc_SA_S610_RT_29_s24                   1  \n",
      "20230516_EXPL1_LiSc_SA_S610_RT_30_s30                   1  \n",
      "20230516_EXPL1_LiSc_SA_S610_RT_31_s33                   1  \n",
      "20230516_EXPL1_LiSc_SA_S610_RT_32_s38                   1  \n",
      "20230516_EXPL1_LiSc_SA_S610_RT_33_s44                   1  \n",
      "20230516_EXPL1_LiSc_SA_S610_RT_34_s48                   1  \n",
      "20230516_EXPL1_LiSc_SA_S610_RT_35_s51                   1  \n",
      "20230516_EXPL1_LiSc_SA_S610_RT_36_s56                   1  \n",
      "20230516_EXPL1_LiSc_SA_S610_RT_37_s63                   1  \n",
      "20230516_EXPL1_LiSc_SA_S610_RT_38_s68                   1  \n",
      "20230516_EXPL1_LiSc_SA_S610_RT_39_s72                   1  \n",
      "20230516_EXPL1_LiSc_SA_S610_RT_40_s80                   1  \n",
      "20230516_EXPL1_LiSc_SA_S610_RT_41_s81                   1  \n",
      "20230516_EXPL1_LiSc_SA_S610_RT_42_s89                   1  \n",
      "20230516_EXPL1_LiSc_SA_S610_RT_43_s92                   1  \n",
      "20230516_EXPL1_LiSc_SA_S610_RT_44_s96                   1  \n",
      "20230516_EXPL1_LiSc_SA_S610_RT_45_QC1_LabD              1  \n",
      "20230516_EXPL1_LiSc_SA_S610_RT_46_QC2_LabD              1  \n",
      "20230516_EXPL1_LiSc_SA_S610_RT_47_QC3_LabD              1  \n",
      "20230516_EXPL1_LiSc_SA_S610_RT_48_QC4_LabD              1  \n",
      "Design matrix shape: (47, 6)\n",
      "Design matrix with summed columns: intercept        47.0\n",
      "Pyr              23.0\n",
      "center1|lab_A     0.0\n",
      "center1|lab_B     0.0\n",
      "center2|lab_C    23.0\n",
      "center2|lab_D    24.0\n",
      "dtype: float64\n",
      "Client center3: Inputs validated.\n",
      "Number of features available in this client: 2363\n",
      "Number of features given globally: 3034\n",
      "Number of features only available on this client: 0\n",
      "Number of features available in other clients but not this client: 671\n",
      "Adding 671 extra global features\n",
      "INFO: Client center3 has only one batch\n",
      "INFO: Client center3 is the reference batch\n",
      "Design matrix created:\n",
      "                     intercept  Pyr  center1|lab_A  center1|lab_B  \\\n",
      "file                                                                \n",
      "CVT09_QC1_LabE_X023        1.0    0             -1             -1   \n",
      "CVT09_QC2_LabE_X024        1.0    1             -1             -1   \n",
      "CVT09_QC3_LabE_X025        1.0    1             -1             -1   \n",
      "CVT09_QC4_LabE_X026        1.0    0             -1             -1   \n",
      "CVT09_s3_X002              1.0    1             -1             -1   \n",
      "CVT09_s10_X003             1.0    1             -1             -1   \n",
      "CVT09_s13_X004             1.0    1             -1             -1   \n",
      "CVT09_s16_X005             1.0    1             -1             -1   \n",
      "CVT09_s25_X006             1.0    1             -1             -1   \n",
      "CVT09_s28_X007             1.0    1             -1             -1   \n",
      "CVT09_s31_X008             1.0    1             -1             -1   \n",
      "CVT09_s36_X009             1.0    1             -1             -1   \n",
      "CVT09_s41_X010             1.0    1             -1             -1   \n",
      "CVT09_s50_X011             1.0    1             -1             -1   \n",
      "CVT09_s55_X012             1.0    0             -1             -1   \n",
      "CVT09_s59_X013             1.0    0             -1             -1   \n",
      "CVT09_s61_X014             1.0    0             -1             -1   \n",
      "CVT09_s67_X015             1.0    0             -1             -1   \n",
      "CVT09_s73_X016             1.0    0             -1             -1   \n",
      "CVT09_s77_X017             1.0    0             -1             -1   \n",
      "CVT09_s84_X018             1.0    0             -1             -1   \n",
      "CVT09_s90_X019             1.0    0             -1             -1   \n",
      "CVT09_s93_X020             1.0    0             -1             -1   \n",
      "CVT09_s99_X021             1.0    0             -1             -1   \n",
      "\n",
      "                     center2|lab_C  center2|lab_D  \n",
      "file                                               \n",
      "CVT09_QC1_LabE_X023             -1             -1  \n",
      "CVT09_QC2_LabE_X024             -1             -1  \n",
      "CVT09_QC3_LabE_X025             -1             -1  \n",
      "CVT09_QC4_LabE_X026             -1             -1  \n",
      "CVT09_s3_X002                   -1             -1  \n",
      "CVT09_s10_X003                  -1             -1  \n",
      "CVT09_s13_X004                  -1             -1  \n",
      "CVT09_s16_X005                  -1             -1  \n",
      "CVT09_s25_X006                  -1             -1  \n",
      "CVT09_s28_X007                  -1             -1  \n",
      "CVT09_s31_X008                  -1             -1  \n",
      "CVT09_s36_X009                  -1             -1  \n",
      "CVT09_s41_X010                  -1             -1  \n",
      "CVT09_s50_X011                  -1             -1  \n",
      "CVT09_s55_X012                  -1             -1  \n",
      "CVT09_s59_X013                  -1             -1  \n",
      "CVT09_s61_X014                  -1             -1  \n",
      "CVT09_s67_X015                  -1             -1  \n",
      "CVT09_s73_X016                  -1             -1  \n",
      "CVT09_s77_X017                  -1             -1  \n",
      "CVT09_s84_X018                  -1             -1  \n",
      "CVT09_s90_X019                  -1             -1  \n",
      "CVT09_s93_X020                  -1             -1  \n",
      "CVT09_s99_X021                  -1             -1  \n",
      "Design matrix shape: (24, 6)\n",
      "Design matrix with summed columns: intercept        24.0\n",
      "Pyr              12.0\n",
      "center1|lab_A   -24.0\n",
      "center1|lab_B   -24.0\n",
      "center2|lab_C   -24.0\n",
      "center2|lab_D   -24.0\n",
      "dtype: float64\n",
      "Index(['intercept', 'Pyr', 'center1|lab_A', 'center1|lab_B', 'center2|lab_C',\n",
      "       'center2|lab_D'],\n",
      "      dtype='object')\n",
      "Index(['intercept', 'Pyr', 'center1|lab_A', 'center1|lab_B', 'center2|lab_C',\n",
      "       'center2|lab_D'],\n",
      "      dtype='object')\n",
      "Index(['intercept', 'Pyr', 'center1|lab_A', 'center1|lab_B', 'center2|lab_C',\n",
      "       'center2|lab_D'],\n",
      "      dtype='object')\n"
     ]
    }
   ],
   "source": [
    "### SIMULATION: All: validate ###\n",
    "### Expand data to fullfill the global format. Also performs consistency checks\n",
    "cols = []\n",
    "for clientWrapper in clientWrappers:\n",
    "\n",
    "    time_tracker[clientWrapper.id] = time.time()\n",
    "\n",
    "    global_feauture_names_hashed, global_variables_hashed, cohorts_order = \\\n",
    "        broadcast_features_variables\n",
    "    client = clientWrapper.client_class\n",
    "    client.validate_inputs(global_variables_hashed)\n",
    "    client.set_data(global_feauture_names_hashed)\n",
    "\n",
    "    err = client.create_design(cohorts_order)\n",
    "    cols.append(client.design.columns)\n",
    "        # we choose the last batch of the last cohort as reference batch\n",
    "    if err:\n",
    "        raise ValueError(err)\n",
    "\n",
    "    end_time = time.time()\n",
    "    time_tracker[clientWrapper.id] = end_time - time_tracker[clientWrapper.id]\n",
    "for col in cols:\n",
    "    print(col)\n",
    "    assert cols[0].equals(col)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO: Global mask true count:\n",
      "1844\n",
      "Shape of global mask: (3034, 6)\n",
      "Head of mask: [[False False False False False  True]\n",
      " [False False  True  True  True  True]\n",
      " [False False  True False  True  True]\n",
      " [False False False False False False]\n",
      " [False False False False False  True]]\n"
     ]
    }
   ],
   "source": [
    "### Simulatuion: Coordinator: create design mask based on feature presence matrix ###\n",
    "### Create the mask for the design matrix based on the feature presence matrix\n",
    "### that will be used for the beta computation\n",
    "for clientWrapper in clientWrappers:\n",
    "    if clientWrapper.is_coordinator:\n",
    "        start_time = time.time()\n",
    "        client = clientWrapper.client_class\n",
    "\n",
    "        n=len(client.feature_names)\n",
    "        k=client.design.shape[1]\n",
    "\n",
    "        global_mask = create_beta_mask(feature_presence_matrix, n, k)\n",
    "        print(f\"Shape of global mask: {global_mask.shape}\")\n",
    "        print(f\"Head of mask: {global_mask[:5]}\")\n",
    "        # memo the global mask\n",
    "\n",
    "        end_time = time.time()\n",
    "        time_tracker[\"Coordinator\"] += end_time - start_time\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "### SIMULATION: All: prepare for compute_XtX_XtY ###\n",
    "\n",
    "for clientWrapper in clientWrappers:\n",
    "    start_time = time.time()\n",
    "\n",
    "    client = clientWrapper.client_class\n",
    "    client.sample_names = client.design.index.values\n",
    "\n",
    "    # Error check if the design index and the data index are the same\n",
    "    # we check by comparing the sorted indexes\n",
    "    client._check_consistency_designfile()\n",
    "\n",
    "    # Extract only relevant (the global) features and samples\n",
    "    client.data = client.data.loc[client.feature_names, client.sample_names]\n",
    "    client.n_samples = len(client.sample_names)\n",
    "\n",
    "    end_time = time.time()\n",
    "    time_tracker[clientWrapper.id] += end_time - start_time"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "### SIMULATION: All: compute_XtX_XtY ###\n",
    "### Compute XtX and XtY and share it\n",
    "send_XtX_XtY_list: List[List[np.ndarray]] = list()\n",
    "for clientWrapper in clientWrappers:\n",
    "    start_time = time.time()\n",
    "\n",
    "    client = clientWrapper.client_class\n",
    "\n",
    "    # compute XtX and XtY\n",
    "    XtX, XtY, err = client.compute_XtX_XtY()\n",
    "    if err != \"\":\n",
    "        raise ValueError(err)\n",
    "\n",
    "    # send XtX and XtY\n",
    "    send_XtX_XtY_list.append([XtX, XtY])\n",
    "\n",
    "    end_time = time.time()\n",
    "    time_tracker[clientWrapper.id] += end_time - start_time"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO: Random XtX entry: [[ 95.  47.   0. -24.  -1.   0.]\n",
      " [ 47.  47.   0. -12.  -1.   0.]\n",
      " [  0.   0.  48.  24.  24.  24.]\n",
      " [-24. -12.  24.  24.  24.  24.]\n",
      " [ -1.  -1.  24.  24.  47.  24.]\n",
      " [  0.   0.  24.  24.  24.  48.]]\n",
      "INFO: All XtX entries are the same?: False\n",
      "INFO: Shape of beta: (3034, 6)\n",
      "INFO: Number of pseudo inverses: 0\n"
     ]
    }
   ],
   "source": [
    "### SIMULATION: Coordinator: compute_beta\n",
    "### Compute the beta values and broadcast them to the others\n",
    "broadcast_betas = None # np.ndarray of shape num_features x design_columns\n",
    "\n",
    "for clientWrapper in clientWrappers:\n",
    "    if clientWrapper.is_coordinator:\n",
    "\n",
    "        start_time = time.time()\n",
    "\n",
    "        client = clientWrapper.client_class\n",
    "        beta = compute_beta(XtX_XtY_list=send_XtX_XtY_list,\n",
    "                            n=len(client.feature_names),\n",
    "                            k=client.design.shape[1],\n",
    "                            global_mask=global_mask)\n",
    "\n",
    "        # send beta to clients so they can correct their data\n",
    "        broadcast_betas = beta\n",
    "\n",
    "        end_time = time.time()\n",
    "        time_tracker[\"Coordinator\"] += end_time - start_time"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Start remove_batch_effects\n",
      "Shape of corrected data after correction: (3034, 47)\n",
      "Amount of index found in hash2feature: 2910/2910\n",
      "remove batch final corrected data shape: (2910, 47)\n",
      "DEBUG: Shape of corrected data: (2910, 47)\n",
      "Start remove_batch_effects\n",
      "Shape of corrected data after correction: (3034, 47)\n",
      "Amount of index found in hash2feature: 2900/2900\n",
      "remove batch final corrected data shape: (2900, 47)\n",
      "DEBUG: Shape of corrected data: (2900, 47)\n",
      "Start remove_batch_effects\n",
      "Shape of corrected data after correction: (3034, 24)\n",
      "Amount of index found in hash2feature: 2363/2363\n",
      "remove batch final corrected data shape: (2363, 24)\n",
      "DEBUG: Shape of corrected data: (2363, 24)\n"
     ]
    }
   ],
   "source": [
    "### SIMULATION: All: include_correction\n",
    "### Corrects the individual data\n",
    "for clientWrapper in clientWrappers:\n",
    "\n",
    "    start_time = time.time()\n",
    "\n",
    "    client = clientWrapper.client_class\n",
    "\n",
    "    # remove the batch effects in own data and safe the results\n",
    "    client.remove_batch_effects(beta)\n",
    "    print(f\"DEBUG: Shape of corrected data: {client.data_corrected.shape}\")\n",
    "\n",
    "    end_time = time.time()\n",
    "    time_tracker[clientWrapper.id] += end_time - start_time\n",
    "\n",
    "    # As this is a simulation we don't save the corrected data to csv, instead\n",
    "    # we save it as a variable to the clientwrapper\n",
    "    clientWrapper.data_corrected = client.data_corrected\n",
    "    # client.data_corrected.to_csv(os.path.join(os.getcwd(), \"mnt\", \"output\", \"only_batch_corrected_data.csv\"),\n",
    "    #                                 sep=self.load(\"separator\"))\n",
    "    # client.data_corrected_and_raw.to_csv(os.path.join(os.getcwd(), \"mnt\", \"output\", \"all_data.csv\"),\n",
    "    #                              sep=self.load(\"separator\"))\n",
    "    # with open(os.path.join(os.getcwd(), \"mnt\", \"output\", \"report.txt\"), \"w\") as f:\n",
    "    #     f.write(client.report)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Time tracker for coordinator, ms: 203.23\n",
      "Time tracker for center1, ms: 235.53\n",
      "Time tracker for center2, ms: 224.88\n",
      "Time tracker for center3, ms: 265.94\n"
     ]
    }
   ],
   "source": [
    "# print the time tracker for the coordinator\n",
    "print(f\"Time tracker for coordinator, ms: {round(time_tracker['Coordinator']*1000, 2)}\")\n",
    "\n",
    "# print the time tracker for the clients\n",
    "for clientWrapper in clientWrappers:\n",
    "    print(f\"Time tracker for {clientWrapper.id}, ms: {round(time_tracker[clientWrapper.id]*1000, 2)}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "###                                  INFO                                  ###\n",
    "###                            SIMULATION IS DONE                          ###\n",
    "### The simulation is done. The corrected data is saved in the             ###\n",
    "### clientWrapper instances. Now we analyse the data by comparing to the   ###\n",
    "### calculated centralized corrected data.                                 ###\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "federated_df, intersect_features = _concat_federated_results(clientWrappers, samples_in_columns=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "### SAVE THE RESULTS ###\n",
    "#federated_df.to_csv(os.path.join(\"..\", \"evaluation_data\", datafolder, \"after\", \"FedSim_corrected_data.tsv\"), sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "_________________________Analysing: proteomics_multibatch_________________________\n",
      "Max difference: 1.5631940186722204e-13\n",
      "Mean difference: 3.16976089953702e-14\n",
      "Max diff at position: 20230516_EXPL1_LiSc_SA_S610_RT_40_s80\n",
      "Max difference in intersect: 1.1723955140041653e-13\n",
      "Mean difference in intersect: 3.33913166227243e-14\n",
      "Max diff at position in intersect: CVT09_s13_X004\n"
     ]
    }
   ],
   "source": [
    "### Concat the federated data and read in the centralized data ###\n",
    "base_dir = os.path.join(\"..\", \"evaluation_data\")\n",
    "central_df_path = os.path.join(base_dir, datafolder, \"after\", central_filename)\n",
    "central_df = pd.read_csv(central_df_path, sep=\"\\t\", index_col=0)\n",
    "_compare_central_federated_dfs(output_name, central_df, federated_df, intersect_features)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base-omRVbgZA-py3.12",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
